src/libraries/System.Private.DataContractSerialization/src/System/Xml/XmlBinaryWriter.cs

```diff
@@ -72,6 +75,12 @@ private void WriteTextNode(XmlBinaryNodeType nodeType)
             _textNodeOffset = this.BufferOffset - 1;
         }
 
+        private ref byte GetBufferRef(int size)
```

Comment:
Yes there was quite much unsafe code and points present in the file so I assumed it was ok to be unsafe for the best performance sine GetBuffer does the bounds check.

I've refactored the code so that all new "unsafe" code is now only present in 2 methods to make it much easier to reason about and updated the benchmark with timings for both Unsafe and Span. 

The span version does  `.AsSpan(offset, size)` on the buffer and then uses `MemoryMarshal.Write<T>` which I just foud out about.. I hope that method is considered safe enough and works well with "unaligned writes".

Please have a look and let me know based on the new code and the benchmark results which version you want.
I can push the updated span version instead if that version is the prefered one.

Reply:
Note: A side effect of moving to the span version is that it will probably be necessary to keep separate implementationa for writing 8bit (and maybe 16bit) nodes in order to not regress performance for them.

Some benchmarks would be required to determine the performance numbers for smaller elements first